R/tm_g_waterfall.R

```diff
@@ -293,15 +293,83 @@ srv_g_waterfall <- function(id,
 
   moduleServer(id, function(input, output, session) {
     output_q <- reactive({
-      iv <- shinyvalidate::InputValidator$new()
-      iv$add_rule("bar_var", shinyvalidate::sv_required())
-      iv$add_rule("bar_paramcd", shinyvalidate::sv_required())
-      iv$enable()
-
       adsl <- data[["ADSL"]]()
       adtr <- data[[dataname_tr]]()
       adrs <- data[[dataname_rs]]()
 
+      # validate data rows
+      teal::validate_has_data(adsl, min_nrow = 2)
+      teal::validate_has_data(adtr, min_nrow = 2)
+      teal::validate_has_data(adrs, min_nrow = 2)
+
+      adsl_vars <- unique(
+        c(
+          "USUBJID", "STUDYID",
+          input$bar_color_var, input$sort_var, input$add_label_var_sl, input$anno_txt_var_sl, input$facet_var
+        )
+      )
+      adtr_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD", input$bar_var))
+      adrs_vars <- unique(c("USUBJID", "STUDYID", "PARAMCD", "AVALC"))
+      adrs_paramcd <- unique(c(input$add_label_paramcd_rs, input$anno_txt_paramcd_rs))
+
+      # validate data input
+      teal::validate_has_variable(adsl, adsl_vars)
+      teal::validate_has_variable(adrs, adrs_vars)
+      teal::validate_has_variable(adtr, adtr_vars)
+
+      iv <- shinyvalidate::InputValidator$new()
+      iv$add_rule("bar_var", shinyvalidate::sv_required(
+        message = "Bar Height is required"
+      ))
+      iv$add_rule("bar_paramcd", shinyvalidate::sv_required(
+        message = "Tumor Burden Parameter is required"
+      ))
+      iv$add_rule("bar_paramcd", shinyvalidate::sv_in_set(
+        set = adtr$PARAMCD,
+        message_fmt = "All values of Tumor Burden Parameter must be elements of ADTR PARAMCD"
+      ))
+      iv$add_rule("href_line", ~ if (anyNA(as_numeric_from_comma_sep_str(.))) {
+        "Horizontal Reference Line(s) are invalid"
+      })
+      iv$add_rule("add_label_paramcd_rs", shinyvalidate::sv_optional())
+      iv$add_rule("add_label_paramcd_rs", shinyvalidate::sv_in_set(
+        set = adrs$PARAMCD,
+        message_fmt = "ADRS Label must be an element of ADRS PARAMCD"
+      ))
+      rule_excl <- function(value, other) {
+        if (length(value) > 0L & length(other) > 0L) {
```

Comment:
```suggestion
        if (length(value) > 0L && length(other) > 0L) {
```

Reply:
This is only wrong if __both__ are length > 0.

Reply:
`length(value) > 0` and `length(other) > 0` both return a single logical value? 

Then you should use "&&" not "&"? as single "&" is the vectorized version of "&&"

![image](https://user-images.githubusercontent.com/15201933/206443537-fa28bc52-2e21-4e60-8e8f-af969d54d8cd.png)